### PR TITLE
Fix data table

### PR DESCRIPTION
**Pull request by N00bG1rl**
## Summary

This PR will fix errors with microservices overview data table, when data is undefined.

Screenshot 2022-08-12 at 17 49 44

### Changed

- Refactored customUrlFieldSort function

### Fixed

- Microservices data table errors

### CODE REVIEW

File: Source/SelfService/Web/microservice/dataTable.tsx [hunk 1]
```diff
@@ -28,25 +28,34 @@ const styles = {
...
+const tooltipCell = (params: GridRenderCellParams) => {
+    const publicUrl = params.row.edit?.extra?.isPublic;
+    if (publicUrl === undefined) return 'N/A';
+
+    return (
+        // TODO: Tooltip needs public urls. Map them into title.
+        <Tooltip title={``} arrow>
+            <span>{params.row.edit?.extra?.isPublic ? 'Available' : 'None'}</span>
+        </Tooltip>
+    );
```
Comment (jakhog) [inline Source/SelfService/Web/microservice/dataTable.tsx:47]: If we return just a string here, the styling might be different. I believe it is better to just change the text inside the tooltip and span:
```suggestion
    const hasPublicUrl = params.row.edit?.extra?.isPublic;

    return (
        // TODO: Tooltip needs public urls. Map them into title.
        <Tooltip title={``} arrow>
            <span>{
                hasPublicUrl === true ? 'Available' :
                hasPublicUrl === false ? 'None' :
                'N/A'
            }</span>
        </Tooltip>
    );
```

File: Source/SelfService/Web/microservice/dataTable.tsx [hunk 2]
```diff
@@ -28,25 +28,34 @@ const styles = {
...
     return `${runtimeVersion || 'None'}`;
 };
 
-const tooltipCell = (params: GridRenderCellParams) => (
-    // TODO: Tooltip needs public urls. Map them into title.
-    <Tooltip title={``} arrow>
-        <span>{params.row.edit.extra.isPublic ? 'Available' : 'None'}</span>
-    </Tooltip>
-);
+const tooltipCell = (params: GridRenderCellParams) => {
```
Comment (jakhog) [inline Source/SelfService/Web/microservice/dataTable.tsx:38]: Lets rename this to:
```suggestion
const publicUrlCell = (params: GridRenderCellParams) => {
```

File: Source/SelfService/Web/microservice/dataTable.tsx [hunk 3]
```diff
@@ -165,7 +175,7 @@ export const DataTable = ({ application, environment, microservices }: DataTable
             flex: 1,
             renderCell: statusCell,
             sortComparator: (v1, v2, param1, param2) =>
-                param1.value[0].phase.localeCompare(param2.value[0].phase)
+                param1.value[0]?.phase?.localeCompare(param2.value[0]?.phase)
```
Comment (jakhog) [inline Source/SelfService/Web/microservice/dataTable.tsx:178]: I believe this function might return `undefined` - which can cause undefined behaviour. So I suggest that we change this a little bit.

First, we can define a set of statuses that we care about - and then return them from the backend status:
```typescript
enum MicroserviceStatus {
    Running = 0,
    Pending = 1,
    Failing = 2,
    Unknown = 3,
};

const getMicroserviceState = (phase?: string): MicroserviceStatus => {
    const checkStatus = phase?.toLowerCase?.();

    if (typeof checkStatus !== 'string') {
        return MicroserviceStatus.Unknown;
    } else if (checkStatus.includes('running')) {
        return MicroserviceStatus.Running;
    } else if (checkStatus.includes('pending')) {
        return MicroserviceStatus.Pending;
    } else if (checkStatus.includes('failed')) {
        return MicroserviceStatus.Failing;
    }

    return MicroserviceStatus.Unknown;
}
```

Then we can change the `statusCell` to use that status for rendering:
```typescript

const statusCell = (params: GridRenderCellParams) => {
    let color = themeDark.palette.text.primary;
    let icon = <QuestionMark sx={{ color }} />;

    switch (getMicroserviceState(params.row.status[0]?.phase)) {
        case MicroserviceStatus.Running:
            icon = <CheckCircleRounded />;
            break;
        case MicroserviceStatus.Pending:
            color = themeDark.palette.warning.main;
            icon = <WarningRounded sx={{ color }} />;
            break;
        case MicroserviceStatus.Failing:
            color = themeDark.palette.error.main;
            icon = <ErrorRounded sx={{ color }} />;
            break;
        case MicroserviceStatus.Unknown:
            return 'N/A';
    }

    return (
        <Box sx={styles.status}>
            {icon}
            <Typography sx={{ ...styles.statusTitle, color }}>{status}</Typography>
        </Box>
    );
}
```

Then, lastly - we can change this sortComparator to sort using the numeric value of that enum we created:
```typescript
            sortComparator: (_, __, left, right) => {
                const leftStatus = getMicroserviceState(left.value[0]?.phase);
                const rightStatus = getMicroserviceState(right.value[0]?.phase);
                return leftStatus - rightStatus;
            }
```

This should result in a stable sorting. If you want to change the order of which things are sorted (maybe we want the errors on top?) - you can just rearrange the numbers in this enum:
```typescript
enum MicroserviceStatus {
    Running = 0,
    Pending = 1,
    Failing = 2,
    Unknown = 3,
};
```

**Review by jakhog — CHANGES_REQUESTED** [review]
Some small suggestions - the big one relating to sorting on the status.

File: Source/SelfService/Web/microservice/dataTable.tsx [hunk 4]
```diff
@@ -28,25 +28,34 @@ const styles = {
     }
 };
 
-export const sortByRuntimeVersion = (params: GridValueGetterParams) => {
-    const runtimeVersion = params.row.edit.extra.runtimeImage.replace(/dolittle\/runtime:/gi, '');
+const sortByRuntimeVersion = (params: GridValueGetterParams) => {
+    const runtimeVersion = params.row.edit?.extra?.runtimeImage?.replace(/dolittle\/runtime:/gi, '');
+    if (runtimeVersion === undefined) return 'N/A';
+
     return `${runtimeVersion || 'None'}`;
```
[the remark this answers is not in the document]
Comment (N00bG1rl) [reply Source/SelfService/Web/microservice/dataTable.tsx:35]: Well, it still needs to return 'None' if there is no runtime version. I could do something like that: runtimeVersion.replace(/^\w/, (c) => c.toUpperCase());

Would that be better solution?

[the remark this answers is not in the document]
Comment (jakhog) [reply Source/SelfService/Web/microservice/dataTable.tsx]: Hmm, I don't understand this. If there is no runtime version - wouldn't the variable `runtimeVersion` be `undefined`? Or am I missing something here..?

**Review by jakhog — APPROVED** [review]

**Pull request merged**